crates/runtime/src/instance/allocator/pooling/memory_pool.rs

```diff
@@ -0,0 +1,439 @@
+
+        let mut slot = self.take_memory_image_slot(allocation_index);
+        let image = request.runtime_info.memory_image(memory_index)?;
+        let initial_size = memory_plan.memory.minimum * WASM_PAGE_SIZE as u64;
+
+        // If instantiation fails, we can propagate the error
+        // upward and drop the slot. This will cause the Drop
+        // handler to attempt to map the range with PROT_NONE
+        // memory, to reserve the space while releasing any
+        // stale mappings. The next use of this slot will then
+        // create a new slot that will try to map over
+        // this, returning errors as well if the mapping
+        // errors persist. The unmap-on-drop is best effort;
+        // if it fails, then we can still soundly continue
+        // using the rest of the pool and allowing the rest of
+        // the process to continue, because we never perform a
+        // mmap that would leave an open space for someone
+        // else to come in and map something.
+        slot.instantiate(initial_size as usize, image, memory_plan)?;
+
+        let memory = Memory::new_static(
+            memory_plan,
+            base_ptr,
+            base_capacity,
+            slot,
+            self.memory_and_guard_size,
+            unsafe { &mut *request.store.get().unwrap() },
+        )?;
+
+        Ok((allocation_index, memory))
+    }
+
+    /// Deallocate a previously-allocated memory.
+    ///
+    /// # Safety
+    ///
+    /// The memory must have been previously allocated from this pool and
+    /// assigned the given index, must currently be in an allocated state, and
+    /// must never be used again.
+    pub unsafe fn deallocate(&self, allocation_index: MemoryAllocationIndex, memory: Memory) {
+        let mut image = memory.unwrap_static_image();
+
+        // Reset the image slot. If there is any error clearing the
+        // image, just drop it here, and let the drop handler for the
+        // slot unmap in a way that retains the address space
+        // reservation.
+        if image.clear_and_remain_ready(self.keep_resident).is_ok() {
+            self.return_memory_image_slot(allocation_index, image);
+        }
+
+        self.index_allocator.free(SlotId(allocation_index.0));
+    }
+
+    /// Purging everything related to `module`.
+    pub fn purge_module(&self, module: CompiledModuleId) {
+        // This primarily means clearing out all of its memory images present in
+        // the virtual address space. Go through the index allocator for slots
+        // affine to `module` and reset them, freeing up the index when we're
+        // done.
+        //
+        // Note that this is only called when the specified `module` won't be
+        // allocated further (the module is being dropped) so this shouldn't hit
+        // any sort of infinite loop since this should be the final operation
+        // working with `module`.
+        for i in 0..self.memories_per_instance {
```

Comment:
leave a comment that we could keep track of how many memories we have per module instead of probing all possible indices if this ever shows up in a profile